src/main/java/emu/grasscutter/command/handler/HandlerContext.java

```diff
@@ -0,0 +1,111 @@
+package emu.grasscutter.command.handler;
+
+import emu.grasscutter.command.handler.exception.NoSuchArgumentException;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.Singular;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.SortedMap;
+import java.util.function.Consumer;
+
+@Getter
+@Builder(toBuilder = true)
+public final class HandlerContext {
+
+    @NotNull
+    @Singular("content")
+    private SortedMap<String, Object> content;
+    @Nullable
+    private Consumer<Object> messageConsumer;
+    @Nullable
+    private Consumer<Object> resultConsumer;
+    @Nullable
+    private Consumer<Exception> errorConsumer;
+
+    /**
+     * Get a required argument from the context.
+     * @param name argument name
+     * @param clazz type of the argument
+     * @return the argument if it exists
+     * @throws NoSuchElementException if argument associated with such a key doesn't exist
+     * @throws IllegalArgumentException if casting to given type failed
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getRequired(@NotNull String name, Class<T> clazz)
+            throws NoSuchElementException, IllegalArgumentException {
+        // no such data
+        Object result = content.get(name);
+        if (result == null) {
+            throw new NoSuchArgumentException(name, clazz);
+        }
+
+        // convert to given type
+        if (Primitive2BoxedType.getOrDefault(clazz, clazz).isAssignableFrom(result.getClass())) {
+            return (T) result;
+        }
+        // cast failed
+        throw new NoSuchArgumentException(name, clazz);
+    }
+
+    /**
+     * Get an optional argument from the context.
+     * @param name argument name
+     * @param defaultValue default value of this argument
+     * @return the argument, or the default value if such argument is not found
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getOptional(@NotNull String name, T defaultValue) throws IllegalArgumentException {
+        try {
+            return (T) getRequired(name, defaultValue.getClass());
+        } catch (NoSuchArgumentException ignored) {
+            return defaultValue;
+        }
+    }
+
+    /**
+     * <p>Pass the result so that others can display.</p>
+     * <b>Notice: </b><p>Code under this will NOT be executed.</p>
+     * @param result the result of this handler
+     */
+    public void returnWith(Object result) throws HandlerSuccess {
+        throw new HandlerSuccess(result);
```

Comment:
I don't understand the purpose here, why throw `'Exception'` of `HandlerSuccess`?

Reply:
`HandlerSuccess` is not an exception. It just interrupts the code and gets caught in `DispatchListener#dispatch`. So that the consumer can be acknowledged that the handler is DONE and you can dispose resources in all callbacks

Reply:
I noticed that your example usage is called after handler is DONE, what's the difference between letting the method return directly?

Reply:
In my opinion, handlers can return something like "online players". `context#notify` just tells the consumers about the progresses etc. It is ok to just access to `HandlerContext#resultConsumer` an then return. But I don't recommand to do so because I want it to seem like "returning something" in a void method.

Reply:
> The signature of %s must be `void (HandlerContext)`.

Maybe make the signature have a return value?
 `void (HandlerContext)` ->  `object (HandlerContext)`

Or put the returned result into the context.

Because throwing HandleSuccess feels weird...

Reply:
Oh yes. In my previous code the handlers are managed by eventbus so they are forced to be void. Now it can just return.
silly me.